cmd/api/src/api/v2/assetgrouptags.go

```diff
@@ -967,3 +968,92 @@ func (s *Resources) GetAssetGroupTagHistory(response http.ResponseWriter, reques
 
 	s.assetGroupTagHistoryImplementation(response, request, "")
 }
+
+type CertifyMembersRequest struct {
+	MemberIDs []int                         `json:"member_ids"`
+	Action    model.AssetGroupCertification `json:"action"`
+	Note      string                        `json:"note,omitempty"`
+}
+
+func isValidCertType(certType model.AssetGroupCertification) bool {
+	switch certType {
+	case model.AssetGroupCertificationManual, model.AssetGroupCertificationRevoked:
+		return true
+	default:
+		return false
+	}
+}
+
+func nodeShouldBeUpdated(node model.AssetGroupSelectorNodeExpanded, lastProcessedNodeId graph.ID, highestPriority int) bool {
+	if node.Certified != model.AssetGroupCertificationAuto {
+		if node.NodeId != lastProcessedNodeId {
+			return true
+		} else if node.Position == highestPriority {
+			return true
+		}
+	}
+	return false
+}
+
+type setSelectorNodeCertifier interface {
+	UpdateCertificationBySelectorNode(ctx context.Context, input []database.UpdateCertificationBySelectorNodeInput) error
+}
+
+// certifyMembersBySelectorNodes - Note: nodes supplied must be in order of nodeId, followed by position
+func certifyMembersBySelectorNodes(ctx context.Context, selectorNodeCertifier setSelectorNodeCertifier, nodes []model.AssetGroupSelectorNodeExpanded, requestAction model.AssetGroupCertification, userEmail null.String, note null.String) error {
+	var (
+		certificationStatus model.AssetGroupCertification
+		certifiedBy         null.String
+
+		lastProcessedNodeId         graph.ID = 0
+		highestPriorityForGivenNode          = -1
+
+		inputs = []database.UpdateCertificationBySelectorNodeInput{}
+	)
+
+	// Only update the nodes with the highest priority for a given nodeID
+	for _, node := range nodes {
+		certificationStatus = model.AssetGroupCertificationPending
+		certifiedBy = null.String{}
+		if nodeShouldBeUpdated(node, lastProcessedNodeId, highestPriorityForGivenNode) {
+			certificationStatus = requestAction
+			highestPriorityForGivenNode = node.Position
+			certifiedBy = userEmail
+			lastProcessedNodeId = node.NodeId
+			// don't actually update nodes that already match the request action
+			if node.Certified == requestAction {
+				continue
+			}
+		}
+		inputs = append(inputs, database.UpdateCertificationBySelectorNodeInput{AssetGroupTagId: node.AssetGroupTagId, SelectorId: node.SelectorId, CertifiedBy: certifiedBy, CertificationStatus: certificationStatus, NodeId: node.NodeId, NodeName: node.NodeName, Note: note})
```

Comment:
```suggestion
			inputs = append(inputs, database.UpdateCertificationBySelectorNodeInput{AssetGroupTagId: node.AssetGroupTagId, SelectorId: node.SelectorId, CertifiedBy: certifiedBy, CertificationStatus: certificationStatus, NodeId: node.NodeId, NodeName: node.NodeName, Note: note})
		}
```
Should the append be inside the if, so no update happens if the node doesn't need updated? Unless it's intentional to hard-code those dupes to `AssetGroupCertificationPending`

Reply:
Good question! It is intended to set the lower-priority dupes to pending, only the highest-priority selector nodes for a given node ID should be updated, and the other nodes must be set to pending (just in case they were manually certified at a different point in time).

Reply:
Good idea, I updated the godoc to better explain how the dupes are handled 👍